packages/processor/processor.js

```diff
@@ -115,7 +121,17 @@ class Processor {
 
         this._log("file()", id);
 
-        return this._add(id, fs.readFileSync(id, "utf8"));
+        if(!this._pendingFiles.has(id)) {
+            const removePending = () => this._pendingFiles.delete(id);
+
+            const pendingAdd = Promise.resolve(this._loadFile(id))
+                .then((contents) => this._add(id, contents));
+
+            pendingAdd.then(removePending, removePending);
+            this._pendingFiles.set(id, pendingAdd);
+        }
```

Comment:
this cache here is required to get the tests to pass, but i'm not sure it's the correct fix. The underlying issue seems to be when the graph get updated. Without the added async, here the first dependent files get added twice (see API test in processor). I think this is because to immediate calls to `file()` from the outside don't await for each, and so assume the graph is updated by the time the next one is called (even tho the work to resolve tho deps happens async).

Reply:
That sounds like a test problem. Adding files is expected to be async, so of the tests aren't awaiting that they're bad because I was being sloppy.

Reply:
I think you'd have to make the call there, here is the test:

```
it("should wait for dependencies to be processed before composing", async () => {
    const results = await Promise.all([
        processor.file(require.resolve("./specimens/overlapping/entry1.css")),
        processor.file(require.resolve("./specimens/overlapping/entry2.css")),
    ]);

    expect(results.map((result) => result.exports)).toMatchSnapshot();
});
```
It doesn't look obviously wrong, but you could easily make the case that one should wait `file()` calls before calling again, as well as make the case that it should just work

Reply:
I don't fully grok why this wouldn't work as the code is written already, so I'll have to pull this down locally and stare at it for a bit.

Reply:
Ah, got it. 2179e24f70a10b044fc12af318978ad949ab53af